### ISSUE TITLE
Predict precision/recall drop when data get bigger

### ISSUE THREAD

**Issue by qinlingzheng** [creme-ml/creme#354]
Hello, 
I have been applying Creme to couple projects I am working on currently. They have all been working very well so far. 
Recently, I applied some classification models to a different case with very big data. Data has 21 features used as converters, I noticed, when the observation is less than 10 million, the prediction accuracy is good. However, when the observation exceeds 20 million, the precision and recall drops to around 0.5. 

I wrote my own performance evaluation function instead of using progressive_val_score(), which was able to increase the performance of two of the evaluated classifiers by about 20%, the rest of the classifiers considered still have very poor prediction accuracy. If I divided the data into smaller datasets, the algorithms work perfect for each individual set of smaller data, but not with the complete set together. 

I have tried using StandardScaler() and RBFSampler() on the raw data. RBFSampler() shows great improvements over StandardScaler().

I have used sklearn() for batch learning successfully on the same complete set of data, but because of the data size will keep on increasing in the long run, I would much rather prefer to use incremental learning to save memory.

Is there any library or configuration that needs to be modified when dealing with data this size? 

Thank you.
Jasmine

**Comment by raphaelsty** [comment]
Hello Jasmine, 

It is difficult to answer your questions without being able to reproduce your experience. If it's possible to you to share the code of your cream pipeline, perhaps this could help us to find a solution.

In the meantime you could try to replace the ```preprocessing.StandardScaler``` with a ```preprocessing.MinMaxScaler``` to check that your experience is not linked to a sudden change of trend in the variables. 

[MinMaxScaler]( https://creme-ml.github.io/api-reference/preprocessing/MinMaxScaler/)

Raphaël

**Comment by qinlingzheng** [comment]
Thank you Raphaël,

Thanks for the quick response. I use the MinMaxScaler you suggested but with no improvement on the result. Final result gives precision = 0.5 and recall = 0.53. This is just like a blindly guess :(

Please see the code for Logistic Regression classifier bellow. Any suggestion will be great! Thank you.
`
REG = 'Regression'
BINARY_CLF = 'Binary'
MULTI_CLF = 'Multiclass'

# Read and select
def EvaluateModel():
    print("Start Model_evaluation\n")

    class Dataset:

        def __init__(self, n_features, category):
            self.n_features = n_features
            self.category = category

        def __iter__(self):
            raise NotImplementedError

        def take(self, k):
            """Yields the k first (``x``, ``y``) pairs."""
            return itertools.islice(self, k)


    class FileDataset(Dataset):

        def __init__(self, n_samples, n_features, category, **dl_params):
            super().__init__(n_features=n_features, category=category)
            self.n_samples = n_samples
            self.dl_params = dl_params

        def _stream_X_y(self, dir):
            raise NotImplementedError

        @property
        def _ready(self):
            """Whether or not the dataset is ready to be read."""
            if self._remote:
                return os.path.isdir(os.path.join(
                    get_data_home(self.dl_params['data_home']),
                    self.dl_params['name']
                ))
            return True

        def __iter__(self):
            data_dir_path = os.path.dirname(__file__)
            yield from self._stream_X_y(data_dir_path)


    class SingleDev(FileDataset):

        def __init__(self):
            super().__init__(
                n_samples=22_242_630,
                n_features=20 , # only 20 features selected as predictor.
                category=BINARY_CLF
            )
        converters = []
        def _stream_X_y(self, directory):
        
            return stream.iter_csv(
                os.path.join('C:/Users/---.csv'), #load csv file
                target_name='Type',
                converters={'Type':int,
                '1':int,
                '2':int,
                '3':int,
                '4':int,
                '5':int,
                '6':float,
                '7':float,
                '8':float,
                '9':float,
                '10':int,
                '11':int,
                '12':int,
                '13':int,
                '14':int,
                '15':int,
                '16':int,
                '17':int,
                '18':int,
                '19':int,
                '20':int},
                field_size_limit=None
            )

    X_y = SingleDev()
    
    model = (
    preprocessing.MinMaxScaler() |#preprocessing.RBFSampler(seed=3) |# StandardScaler() 
    sampling.RandomUnderSampler(
        classifier=linear_model.LogisticRegression(optimizer=optim.SGD(0.5)),  
        desired_dist={0: 0.5, 1: 0.5},
        seed=42
    ))
    
    ##=================================================================================
    # Using progressive_val_score() ends with precision/recall ~ 0.5
    # skm.classification_report shows different result (precision= 0.9, recall = 0.7) for the same data on the same classifier
    # Training on sub-set of the data shows accuracy > 0.9
    # Batch train result on the same big dataset shows accuracy > 0.9
##=================================================================================
    # metric = metrics.ConfusionMatrix() + metrics.MacroPrecision() + metrics.MacroRecall() + metrics.MacroF1()  

    # model_selection.progressive_val_score(
        # X_y,
        # model,
        # metric,
        # print_every=2_000_000,
        # show_time=True)
        
    # print (metric)
    
    y_trues = []
    y_preds = []

    for x, y in SingleDev():
        y_pred = model.predict_one(x)
        model.fit_one(x, y)
        
        y_trues.append(y)
        y_preds.append(y_pred)

    y_trues = list(y_trues)
    y_preds = list(y_preds)

    cm = skm.multilabel_confusion_matrix(y_trues, y_preds)
    print(cm)
    print(skm.classification_report(y_trues,y_preds))

def main():
    tic = time.perf_counter()
    EvaluateModel()
    toc = time.perf_counter()
    print(f"Training time in {toc - tic:0.4f} seconds")
    
    print("\n Close the Console\n")
    input("\n Press <Enter> to Close...\n")

if __name__ == "__main__":
    main()

`

**Comment by MaxHalford** [comment]
Hey @qinlingzheng! Just to be clear: your issue is that `progressive_val_score` returns different results than `skm.classification_report` when the dataset is large? In other words, it's not a problem with the model, right?

We've made a lot changes to the `metrics` module that aren't released yet, maybe that will help.

**Comment by qinlingzheng** [comment]
Hello @MaxHalford,
Regarding `progressive_val_score` returning different results than `skm.classification_report`, this is one of the issues i have and I am currently using other ways to show results. Thanks for letting me know that there will be new version released soon. 
My biggest issue is even when using `skm.classification_report` why the classifier returns a much more accurate prediction result on a subset of my big dataset, but very poor prediction on the complete dataset?
Thank you very much.

**Comment by MaxHalford** [comment]
I'm not sure, it's really weird... Could you potentially save the `y_trues` and `y_preds` arrays and send them? Either here or by email to [maxhalford25@gmail.com](mailto:maxhalford25@gmail.com).

**Comment by MaxHalford** [comment]
I've taken a look at the files you sent me. Your data is extremely imbalanced (0.0023% of 1s). Are you doing anomaly detection? I looked and it seems that the `1`s are somewhat spread out in time, therefore [HalfSpaceTrees](https://creme-ml.github.io/api-reference/anomaly/HalfSpaceTrees/) might work better for you.

I'm sorry but I can't do more without accessing the training set...

**Comment by qinlingzheng** [comment]
Hello MaxHalford,
I am doing supervised learning currently, not anomaly detection. The supervised batch training results on the same data are very good so far, But our data are very large and will get bigger as time goes on, so incremental learning are explored to solve our case to ease the requirement on the memory. We tried unsupervised learning approach before even when using batch training, however, FNR are too high. Thanks for your suggestions though, will try some other approaches. Best regards,
